File: apps/dsa-web/src/api/shadow.ts
import apiClient from './index';
import { toCamelCase } from './utils';
import type {
  ShadowApprovalRequest,
  ShadowBacktestRequest,
  ShadowBacktestResponse,
  ShadowProfileCreateRequest,
  ShadowProfileDetail,
  ShadowProfileListResponse,
  ShadowObservation,
  ShadowScanRequest,
  ShadowSignalListResponse,
} from '../types/shadow';

function toSnakeObservation(observation: ShadowObservation): Record<string, unknown> {
  return {
    date: observation.date,
    features: observation.features,
    next_return_pct: observation.nextReturnPct,
    features_as_of: observation.featuresAsOf,
    source_refs: observation.sourceRefs,
    evidence_refs: observation.evidenceRefs,
  };
}

function toSnakeBacktestPayload(payload: ShadowBacktestRequest): Record<string, unknown> {
  const request: Record<string, unknown> = {
    code: payload.code.trim().toUpperCase(),
    split_date: payload.splitDate,
    market_data_start: payload.marketDataStart,
    market_data_end: payload.marketDataEnd,
    fee_bps: payload.feeBps,
    slippage_bps: payload.slippageBps,
    source_refs: payload.sourceRefs,
    account_id: payload.accountId,
    ledger_start_date: payload.ledgerStartDate,
    ledger_end_date: payload.ledgerEndDate,
  };
  if (payload.observations?.length) {
    request.observations = payload.observations.map(toSnakeObservation);
  }
  return request;
}

function toSnakeScanPayload(payload: ShadowScanRequest): Record<string, unknown> {
  const request: Record<string, unknown> = {
    code: payload.code.trim().toUpperCase(),
    market_data_start: payload.marketDataStart,
    market_data_end: payload.marketDataEnd,
    run_id: payload.runId,
    evidence_refs: payload.evidenceRefs,
  };
  if (payload.observations?.length) {
    request.observations = payload.observations.map(toSnakeObservation);
  }
  return request;
}

export const shadowApi = {
  async listProfiles(status?: string): Promise<ShadowProfileListResponse> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/shadow/profiles', {
      params: status ? { status } : undefined,
    });
    return toCamelCase<ShadowProfileListResponse>(response.data);
  },

  async createProfile(payload: ShadowProfileCreateRequest): Promise<ShadowProfileDetail> {
    const response = await apiClient.post<Record<string, unknown>>('/api/v1/shadow/profiles', {
      name: payload.name.trim(),
      rule: payload.rule,
      description: payload.description?.trim() || undefined,
      version: payload.version?.trim() || undefined,
    });
    return toCamelCase<ShadowProfileDetail>(response.data);
  },

  async runBacktest(profileId: string, payload: ShadowBacktestRequest): Promise<ShadowBacktestResponse> {
    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/shadow/profiles/${encodeURIComponent(profileId)}/backtest`,
      toSnakeBacktestPayload(payload),
    );
    return toCamelCase<ShadowBacktestResponse>(response.data);
  },

  async approveProfile(profileId: string, payload: ShadowApprovalRequest = {}): Promise<ShadowProfileDetail> {
    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/shadow/profiles/${encodeURIComponent(profileId)}/approve`,
      { approved_by: payload.approvedBy?.trim() || 'local_user' },
    );
    return toCamelCase<ShadowProfileDetail>(response.data);
  },

  async scanSignals(profileId: string, payload: ShadowScanRequest): Promise<ShadowSignalListResponse> {
    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/shadow/profiles/${encodeURIComponent(profileId)}/scan`,
      toSnakeScanPayload(payload),
    );
    return toCamelCase<ShadowSignalListResponse>(response.data);
  },

  async listSignals(profileId: string, code?: string): Promise<ShadowSignalListResponse> {
    const response = await apiClient.get<Record<string, unknown>>(
      `/api/v1/shadow/profiles/${encodeURIComponent(profileId)}/signals`,
      { params: code?.trim() ? { code: code.trim().toUpperCase() } : undefined },
    );
    return toCamelCase<ShadowSignalListResponse>(response.data);
  },
};

File: api/v1/endpoints/shadow.py
# -*- coding: utf-8 -*-
"""Shadow Research API."""

from __future__ import annotations

import logging
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query

from api.deps import get_database_manager
from api.v1.schemas.shadow import (
    ShadowApprovalRequest,
    ShadowBacktestRequest,
    ShadowProfileCreateRequest,
    ShadowScanRequest,
)
from src.shadow_research.repository import ShadowResearchRepository
from src.shadow_research.service import ShadowResearchService, ShadowStateError
from src.storage import DatabaseManager

logger = logging.getLogger(__name__)
router = APIRouter()


def _service(db_manager: DatabaseManager) -> ShadowResearchService:
    return ShadowResearchService(ShadowResearchRepository(db_manager))


@router.post("/profiles")
def create_shadow_profile(request: ShadowProfileCreateRequest, db_manager: DatabaseManager = Depends(get_database_manager)):
    try:
        return _service(db_manager).create_profile(name=request.name, rule=request.rule, description=request.description, version=request.version)
    except (ValueError, KeyError) as exc:
        raise HTTPException(status_code=400, detail={"error": "invalid_shadow_profile", "message": str(exc)})


@router.get("/profiles")
def list_shadow_profiles(status: Optional[str] = Query(None), db_manager: DatabaseManager = Depends(get_database_manager)):
    return {"items": _service(db_manager).list_profiles(status=status)}


@router.get("/profiles/{profile_id}")
def get_shadow_profile(profile_id: str, db_manager: DatabaseManager = Depends(get_database_manager)):
    result = _service(db_manager).get_profile(profile_id)
    if result is None:
        raise HTTPException(status_code=404, detail={"error": "shadow_profile_not_found", "message": "Shadow profile not found"})
    return result


@router.post("/profiles/{profile_id}/backtest")
def run_shadow_backtest(profile_id: str, request: ShadowBacktestRequest, db_manager: DatabaseManager = Depends(get_database_manager)):
    try:
        return _service(db_manager).run_backtest(
            profile_id,
            code=request.code,
            observations=request.observations,
            split_date=request.split_date,
            fee_bps=request.fee_bps,
            slippage_bps=request.slippage_bps,
            source_refs=request.source_refs,
            account_id=request.account_id,
            ledger_start_date=request.ledger_start_date,
            ledger_end_date=request.ledger_end_date,
            market_data_start=request.market_data_start,
            market_data_end=request.market_data_end,
        )
    except ShadowStateError as exc:
        raise HTTPException(status_code=409, detail={"error": "shadow_state_error", "message": str(exc)})
    except ValueError as exc:
        raise HTTPException(status_code=400, detail={"error": "invalid_shadow_backtest", "message": str(exc)})


@router.post("/profiles/{profile_id}/approve")
def approve_shadow_profile(profile_id: str, request: ShadowApprovalRequest, db_manager: DatabaseManager = Depends(get_database_manager)):
    try:
        return _service(db_manager).approve_profile(profile_id, approved_by=request.approved_by)
    except ShadowStateError as exc:
        raise HTTPException(status_code=409, detail={"error": "shadow_state_error", "message": str(exc)})


@router.post("/profiles/{profile_id}/scan")
def scan_shadow_signals(profile_id: str, request: ShadowScanRequest, db_manager: DatabaseManager = Depends(get_database_manager)):
    try:
        return {"items": _service(db_manager).scan_signals(
            profile_id,
            code=request.code,
            observations=request.observations,
            market_data_start=request.market_data_start,
            market_data_end=request.market_data_end,
            run_id=request.run_id,
            evidence_refs=request.evidence_refs,
        )}
    except ShadowStateError as exc:
        raise HTTPException(status_code=409, detail={"error": "shadow_state_error", "message": str(exc)})
    except ValueError as exc:
        raise HTTPException(status_code=400, detail={"error": "invalid_shadow_scan", "message": str(exc)})


@router.get("/profiles/{profile_id}/signals")
def list_shadow_signals(profile_id: str, code: Optional[str] = Query(None), limit: int = Query(100, ge=1, le=500), db_manager: DatabaseManager = Depends(get_database_manager)):
    if _service(db_manager).get_profile(profile_id) is None:
        raise HTTPException(status_code=404, detail={"error": "shadow_profile_not_found", "message": "Shadow profile not found"})
    repository = ShadowResearchRepository(db_manager)
    return {"items": repository.list_signals(profile_id, code=code, limit=limit)}
